datalad/distribution/get.py

```diff
@@ -309,9 +309,75 @@ def _install_subds_from_flexible_source(ds, sm, **kwargs):
                 res.get('path', None) == dest_path:
             _fixup_submodule_dotgit_setup(ds, sm_path)
 
-            # do fancy update
+            target_commit = sm['gitshasum']
             lgr.debug("Update cloned subdataset {0} in parent".format(dest_path))
-            ds.repo.update_submodule(sm_path, init=True)
+            section_name = 'submodule.{}'.format(sm['gitmodule_name'])
+            # do not use `git-submodule update --init`, it would make calls
+            # to git-config which will not obey datalad inter-process locks for
+            # modifying .git/config
+            sub = GitRepo(res['path'])
+            # record what branch we were on right after the clone
+            # TODO instead of the active branch, this should first consider
+            # a configured branch in the submodule record of the superdataset
+            sub_orig_branch = sub.get_active_branch()
+            # if we are on a branch this hexsha will be the tip of that branch
+            sub_orig_hexsha = sub.get_hexsha()
+            # make sure we have the desired commit locally
+            sub.fetch(remote='origin', refspec=target_commit)
```

Comment:
Thanks for the update.

I think we should guard this fetch with a `commit_exists` call because 1) in the common case that commit will exist and that local check is less expensive than the unnecessary fetching and 2) a direct fetch of a commit may fail depending on the remote's configuration

If the commit isn't present locally and the fetch fails, it seems like it'd be good to yield an error result so the operation can respect `--on-failure`.

Reply:
Done. Thx for the analysis. I fully agree.

Reply:
Thanks.  I've tweaked it a bit more to catch the CommandError and added a test.  Of course feel free to drop an of the changes.

Reply:
Thx very much, in particular for the test!